refactor(filetype): prefer explicit filename/extension match over pattern (if possible)

echasnovski opened the pull request:
This PR updates filetype matching by moving as many pattern matches as possible to be direct filename or extension matching. For example:

- Pattern '^\[mM\]akefile.am$' -> two `filename` entries 'makefile.am' and 'Makefile.am'.
- Pattern '%.\[Ss\]\[Yy\]\[Ss\]$' -> eight `extension` entries 'sys', 'Sys', 'sYs', 'SYs', 'syS', 'SyS', 'sYS', 'SYS'.

Moving "filename like" patterns converts 12 patterns into 21 `filename` entries. NOTE: this is not strictly backwards compatible, because exact file name matching is done *before* pattern matching. If user has conflicting `vim.filetype.add()` call with high priority (like `vim.filetype.add({ pattern = { ['.*file%.am'] = { 'my-filetype', { priority = 100 } } } })`), after this change it will be ignored (i.e. 'makefile.am' will match as 'automake' instead of 'my-filetype'). Judging by converted cases, it seems reasonable to prefer exact matches there.

Moving "extension like" patterns converts 12 patterns into 99 `extension` entries. This is backward compatible as all of them have priority 0 making those pattern matching and explicit extension matching equivalent.

------

There are at least two reasons for doing this:

- Performance. It will reduce the number of needed pattern matches (especially moving extensions, as those don't have a "fast" parent pattern). Profiling showed about 5% improvement on cases used in #29660.
- My future plan to suggest adding `fast` option flag to `vim.filetype.match()` which will do only explicit filename and extension matching. This will allow about 10x speed improvement for plugins/users at the cost of not doing pattern and content matching. Having as much explicit filename and extension matches as possible will allow more filetype coverage.

echasnovski committed:
refactor(filetype): use extension match instead of pattern if possible

Problem: some patterns are used as a replacement for several explicit
  extension matches (like '%.[Ss][Yy][Ss]$', '%.php%d$', etc.).
  They usually correspond to Vim's "ignore case" regexes (like
  '*.sys\c') and "convenience" patterns to not define many of them (like
  '*.php\d').
  As matching extension directly is faster and more explicit, it should
  be preferred.
Solution: move all such patterns to direct extension match.

clason commented on runtime/lua/vim/filetype.lua:
```diff
@@ -273,6 +273,16 @@ local extension = {
   hgrc = 'cfg',
+  -- Extension match does not conflict with specific patterns such as '.*/etc/a2ps/.*%.cfg', etc.,
+  -- as it is done after those are tried to match
+  cfg = detect.cfg,
```
can we generate/dispatch that? or add a separate "case insensitive" table?

echasnovski replied:
Both are possible, of course.

I'd say that generation will introduce avoidable complexity only for the sake of shaving off some lines of code.

Separate case insensitive matching sounds better and closer to what is done in Vim's 'filetype.vim'. There are two downsides:
- It won't fit all the cases ('php%d' and '%d' cases will still remain).
- It will add `string.lower()` call and extra table lookup when they can be avoided.

I think my personal preference would be to leave as is in this PR currently.

echasnovski replied:
> I'm just talking about the "full case-insensitive" extension -- my worry is that it's too easy to miss a case (beyond the cumbersomeness). A separate table for these 10 may be a worthwhile tradeoff.

I would agree If not for performance cost: the actual match after extra `string.lower()` and table lookup will is *extremely* rare. So this will essentially add extra avoidable cost for most cases.

And I certainly can see the "cumbersomeness". Especially if there will be something like "php%d%d" with its 100 matches :) But those cases are still can be handled via pattern: the cost will be a slight decrease in performance and (if merged) not working with future proposed `fast` flag.

> At the least, could you add this to the documentation for porting filetype patches? ("Prefer explicit lists over regex/patterns, especially for case insensitive matches".)

Sure, will do. Will adding to 'dev_vimpatch.txt' be enough or should I also add to the comment at the top of `pattern`?

clason replied:
I think we should keep things in one place, so `:h dev_vimpatch`. It doesn't hurt to add a comment pointing to the help entry in a prominent place, too.

> the actual match after extra string.lower() and table lookup will is extremely rare

Well, there's also LPeg...

echasnovski replied:
Added to 'dev_vimpatch.txt' with slightly different wording ("explicit lists over regex/patterns" -> "explicit filenames/extensions over patterns".

echasnovski commented:
> NOTE: this is not strictly backwards compatible, because exact file name matching is done _before_ pattern matching. ... Judging by converted cases, it seems reasonable to prefer exact matches there.

Let me clarify this bit and why I think this change fits the current 'vim.filetype` design. Imagine the following code:

```lua
vim.filetype.add({ pattern = { ['.*%.am'] = { 'my-am', { priority = 100 } } } })

print(vim.inspect({
  gnumake = vim.filetype.match({ filename = 'GNUmakefile.am' }),
  make = vim.filetype.match({ filename = 'Makefile.am' }),
}))
```

Sourcing this on current `master` would print `{ gnumake = 'automake', make = 'my-am' }`.
Sourcing this with this PR would print `{ gnumake = 'automake', make = 'automake' }`.

This is because currently 'Makefile.am' is matched via '^[mM]akefile%.am$' pattern with priority 0, while on this PR it will match by file name exactly (same as currently is done for 'GNUmakefile.am'). Which way the pattern match is done is currently an implementation detail. After this PR it becomes "if the match can be done via exact match, it is done via exact match" for built-in filetypes.

The upstream Vim behavior [doesn't differentiate between these too](https://github.com/vim/vim/blob/37853b7de31ef34153fe76aa2b740d517ed0e5d4/runtime/filetype.vim#L128) and so does current `vim.filetype.match()` without user config. The difference is only visible after user adds conflicting pattern detection with high priority.

echasnovski committed:
docs(filetype): add note about prefering explicit lists over pattern

clason commented on runtime/doc/dev_vimpatch.txt:
```diff
@@ -308,6 +308,9 @@ FILETYPE DETECTION                 *dev-vimpatch-filetype*
 |vim.filetype| (see $VIMRUNTIME/lua/vim/filetype.lua).
 
+Prefer explicit filenames/extensions over patterns, especially for case
+insensitive matches.
```
Maybe expand a little and give examples? This should be enough guidance for new contributors even without any of us around.

clason replied:
(And if you link here, the code comment can be simplified.)

echasnovski replied:
Force pushed some examples while also linking to this PR.

echasnovski committed:
refactor(filetype): use file name match instead of pattern if possible

Problem: some patterns are used as a replacement for one-two explicit
  file matches (like '^[mM]akefile$'). As matching file name directly is
  faster and more explicit, it should be preferred.
Solution: move those patterns to direct file name match.
  NOTE: this is not strictly backwards compatible, because exact file
  name matching is done *before* pattern matching. If user has
  conflicting `vim.filetype.add()` call with high priority (like with
  `pattern='file$'` and `priority=100`), after this change it will be
  ignored (i.e. 'makefile' will match exactly).
  Judging by converted cases, it seems reasonable to prefer exact
  matches there.

echasnovski commented:
Resolved merge conflicts after 60967cd9aa.

jvoisin commented on runtime/lua/vim/filetype.lua:
```diff
@@ -834,6 +901,16 @@ local extension = {
   ctp = 'php',
   php = 'php',
   phpt = 'php',
+  php0 = 'php',
+  php1 = 'php',
```
`php0` and `php1` aren't a thing.

clason replied:
Gotta tell that to Vim; they recognize this and so will we.

clason committed:
docs(filetype): consolidate comments in dev_vimpatch.txt

echasnovski commented on runtime/lua/vim/filetype.lua:
```diff
@@ -174,9 +174,15 @@ end
--- Filetypes based on file extension
+-- Filetype detection logic is encoded in three tables:
+-- 1. `extension` for literal extension lookup
+-- 2. `filename` for literal full filename or path lookup
```
That's not quite right. `filename` is used with both [full path](https://github.com/neovim/neovim/blob/eb629cce917155a4d436327db10c8b1b44ae7861/runtime/lua/vim/filetype.lua#L2689) and [basename](https://github.com/neovim/neovim/blob/eb629cce917155a4d436327db10c8b1b44ae7861/runtime/lua/vim/filetype.lua#L2696).

clason replied:
That's what "full filename" (as opposed to extension) "or path" refers to.

echasnovski commented on runtime/doc/dev_vimpatch.txt:
```diff
@@ -308,36 +308,54 @@ used in new documentation:
+  - specific: rules of thumb, in order:
+    - full directory name (e.g., `"/etc/"`, `"/log/"`);
+    - part of a rare enough directory name (e.g., `"/conf"`, `"git/"`);
+  - string rarely used in real full paths (e.g., `"nginx"`).
```
```suggestion
    - string rarely used in real full paths (e.g., `"nginx"`).
```

echasnovski commented:
@clason, thanks for reorganizing comments! Now looks better than before.

justinmk commented on runtime/lua/vim/filetype.lua:
```diff
@@ -1043,6 +1132,14 @@ local extension = {
   sqi = 'sqr',
   sqr = 'sqr',
   nut = 'squirrel',
+  src = detect.src,
+  srC = detect.src,
+  sRc = detect.src,
+  sRC = detect.src,
+  Src = detect.src,
+  SrC = detect.src,
+  SRc = detect.src,
+  SRC = detect.src,
```
This seems pretty onerous, and reduces readability. We're doing this for a 5% perf improvement?

echasnovski replied:
> This seems pretty onerous, and reduces readability. We're doing this for a 5% perf improvement?

5% is not *that* small improvement. Plus:

> There are at least two reasons for doing this:
> 
> - Performance. It will reduce the number of needed pattern matches (especially moving extensions, as those don't have a "fast" parent pattern). Profiling showed about 5% improvement on cases used in #29660.
> - My future plan to suggest adding `fast` option flag to `vim.filetype.match()` which will do only explicit filename and extension matching. This will allow about 10x speed improvement for plugins/users at the cost of not doing pattern and content matching. Having as much explicit filename and extension matches as possible will allow more filetype coverage.

I had to reschedule a bit my work on `vim.filetype.match()`. I have plans for fixing inconsistency in how `contents` is treated compared to `buf` and adding described `fast` option.

justinmk replied:
The 10x improvement from "fast" doesn't require any of this manual expansion, does it?

echasnovski replied:
> The 10x improvement from `fast` doesn't require any of this manual pattern expansion, does it?

If left as pattern matching, they would not match with `fast` (as it is planned to just skip this phase). With explicit extension match (as after this PR) all those cases ("src", "SRC", etc.) will properly match even with `fast`.

justinmk replied:
I would guess that 80% of the gains from `fast` is that it skips *content* matching. Pattern-matching names, especially simple things like case insensitivity, are unlikely to have a big perf cost. Right?

echasnovski replied:
> I would guess that 80% of the gains from `fast` is that it skips _content_ inspection. Pattern-matching names, especially simple things like case insensitivity, are unlikely to have a big perf cost. Right?

My benchmarking shows otherwise. See #29660.

justinmk replied:
If the caller wants fast, then they don't need to match `SRc` and related variants. We shouldn't have to hardcode case permutations into our list of filenames.

echasnovski replied:
Besides effect on the (already suggested in #30086) `fast` option for `vim.filetype.match()`, reducing the number of times `vim.filetype.match()` does string pattern matching improves performance in the default `fast = false` case. As was already said, by about 5%, which is not quite negligible.

I indeed hesitated to suggest this change, but the current footprint of (12 pattern into 99 extensions) seemed borderline acceptable for the benefits it provides.

Also, adding new case-insensitive extension-like matches is *very* uncommon. All such changes seem to have been done in two commits in April 2022:
- https://github.com/vim/vim/commit/3ad2090316edc85e93094bba7af64f9991cc7f85
- https://github.com/vim/vim/commit/0bbf09ca41382302493e5db51b01d2fbdc778586

With this frequency, I don't think will add too much maintenance burden.
